src/BackgroundTask/CronBasedBackgroundTask.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Abstract class CronBasedBackgroundTask.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\BackgroundTask;
+
+use AmpProject\AmpWP\HasActivation;
+use AmpProject\AmpWP\HasDeactivation;
+use AmpProject\AmpWP\Service;
+
+/**
+ * Abstract base class for using cron to execute a background task.
+ *
+ * @package AmpProject\AmpWP
+ */
+abstract class CronBasedBackgroundTask implements Service, HasDeactivation {
+
+	const DEFAULT_INTERVAL_HOURLY      = 'hourly';
+	const DEFAULT_INTERVAL_TWICE_DAILY = 'twicedaily';
+	const DEFAULT_INTERVAL_DAILY       = 'daily';
+
+	/**
+	 * List of default interval names that are shipped with WordPress.
+	 *
+	 * @var string[]
+	 */
+	const DEFAULT_INTERVAL_NAMES = [
+		self::DEFAULT_INTERVAL_HOURLY,
+		self::DEFAULT_INTERVAL_TWICE_DAILY,
+		self::DEFAULT_INTERVAL_DAILY,
+	];
+
+	/**
+	 * Register the service with the system.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		add_action( 'admin_init', [ $this, 'schedule_event' ] );
+		add_action( $this->get_event_name(), [ $this, 'process' ] );
+	}
+
+	/**
+	 * Schedule the event.
+	 *
+	 * This does nothing if the event is already scheduled.
+	 *
+	 * @return void
+	 */
+	public function schedule_event() {
+		if ( ! current_user_can( 'administrator' ) ) {
+			return;
+		}
+
```

Comment:
Do you think this is required? I think the event should be scheduled if any user logs in.

Also, this is using a role name in the capability check. It could be that a site doesn't have a traditional administrator role, and instead has fine-grained roles that have the capabilities (e.g. `manage_options`)

```suggestion
```

The important thing is `is_admin()` not is _administrator_.

Reply:
That was based on something @schlessera and I discussed. By only running this for administrators and not _any_ users we can reduce the chance for race conditions (there may be massive sites with so many logged in users that that could happen, but there's probably no site in the world with enough administrator users).

I personally have never used this approach, I'd argue without it in the majority of cases there  still won't be a race condition, and for sites where it would, they'd have much bigger problems with WordPress overall.

I'll defer to you guys for which approach to take, but I agree that _if_ this is kept it should use a capability, not a role name (`manage_options` makes sense).